File: ruby/test/document_test.rb
# frozen_string_literal: true

require_relative 'test_helper'

class DocumentTest < Minitest::Test
  def setup
    @mock = MockAdapter.new
    @client = Idkollen::ClientBuilder.new('cid', 'sec').http_client(@mock.connection).build
  end

  def test_upload
    @mock.enqueue(200, { 'id' => 'doc1', 'hash' => 'h1' })
    result = @client.document.upload('PDF-DATA', 'contract.pdf')
    assert_equal 'doc1', result.id
    assert_equal 'h1', result.hash

    req = @mock.last_request
    assert_equal :post, req[:method]
    assert_match %r{\Amultipart/form-data}, req[:headers]['Content-Type']
  end

  def test_download_returns_raw_bytes
    @mock.enqueue(200, "\x25PDF-binary-bytes")
    result = @client.document.download('doc1')
    assert_equal "\x25PDF-binary-bytes", result
  end

  def test_delete
    @mock.enqueue(204, '')
    @client.document.delete('doc1')
    assert_equal :delete, @mock.last_request[:method]
    assert_equal '/document/doc1', @mock.last_request[:path]
  end
end

File: dart/test/document_test.dart
import 'dart:convert';

import 'package:idkollen_client/idkollen_client.dart';
import 'package:test/test.dart';

import 'mock_client.dart';

void main() {
  late MockClient mock;
  late IdkollenClient client;

  setUp(() {
    mock = MockClient();
    client = IdkollenClientBuilder('cid', 'sec').baseUrl('https://x.test').httpClient(mock).build();
  });

  test('upload', () async {
    mock.enqueue(200, {'id': 'doc1', 'hash': 'h1'});
    final data = utf8.encode('PDF-DATA');
    final result = await client.document.upload(data, 'contract.pdf', mimeType: 'application/pdf');
    expect(result.id, 'doc1');
    expect(result.hash, 'h1');

    expect(mock.lastRequest.headers['Content-Type'], startsWith('multipart/form-data'));
    final body = mock.lastRequestBody;
    expect(body, contains('filename="contract.pdf"'));
    expect(body.toLowerCase(), contains('content-type: application/pdf'));
    expect(body, contains('PDF-DATA'));
  });

  test('download', () async {
    final expected = utf8.encode('\x25PDF-binary-bytes');
    mock.enqueue(200, '\x25PDF-binary-bytes');
    final result = await client.document.download('doc1');
    expect(result, expected);
  });

  test('delete', () async {
    mock.enqueue(204, '');
    await client.document.delete('doc1');
    expect(mock.lastRequest.method, 'DELETE');
    expect(mock.lastRequest.url.path, '/document/doc1');
  });
}
